### PR TITLE
fix(forge): `test` result/error handling

### PR DESCRIPTION
**Pull request by DaniPopes**
## Motivation

All `Result::Err` variants are currently ignored (other than an error log) when running any test: https://github.com/foundry-rs/foundry/blob/a5c27411c7e5d1dd574df758060da8ad501ecbde/forge/src/multi_runner.rs#L161

This PR makes methods that execute tests return only the `Ok` type, which includes a `success` or similar field. This heavily simplifies logic around these methods and avoids `unwrap` calls in callers or silent failures in CLI.

### CODE REVIEW

File: forge/src/lib.rs [hunk 1]
```diff
@@ -1,12 +1,13 @@
...
 use foundry_config::{
     validate_profiles, Config, FuzzConfig, InlineConfig, InlineConfigError, InlineConfigParser,
     InvariantConfig, NatSpec,
 };
 use proptest::test_runner::{RngAlgorithm, TestRng, TestRunner};
-use tracing::trace;
+use std::path::Path;
+
+#[macro_use]
+extern crate tracing;
```
Comment (mattsse) [inline forge/src/lib.rs:10]: why do we need this?

File: forge/src/multi_runner.rs [hunk 2]
```diff
@@ -144,48 +143,40 @@ impl MultiContractRunner {
...
     fn run_tests(
         &self,
         name: &str,
         contract: &Abi,
         executor: Executor,
         deploy_code: Bytes,
         libs: &[Bytes],
-        (filter, test_options): (&impl TestFilter, TestOptions),
-    ) -> Result<SuiteResult> {
+        filter: &impl TestFilter,
```
Comment (mattsse) [inline forge/src/multi_runner.rs:177]: -.- `&impl` looks weird lol

File: forge/src/result.rs [hunk 3]
```diff
@@ -184,8 +195,53 @@ pub struct TestSetup {
...
     /// Addresses labeled during setup
     pub labeled_addresses: BTreeMap<Address, String>,
-    /// Whether the setup failed
-    pub setup_failed: bool,
-    /// The reason the setup failed
+    /// The reason the setup failed, if it did
     pub reason: Option<String>,
 }
+
+impl TestSetup {
```
Comment (mattsse) [inline forge/src/result.rs:202]: way better!

**Review by mattsse — APPROVED** [review]
the existing runner code has a lot of baggage -.-

this includes a few improvements

File: forge/src/lib.rs (hunk 1 above)
Comment (DaniPopes) [reply forge/src/lib.rs:10]: less imports and less `tracing::*` when calling, I like it more for macros

File: cli/src/cmd/forge/script/runner.rs [hunk 4]
```diff
@@ -227,7 +226,7 @@ impl ScriptRunner {
 
                     (Address::zero(), gas_used, logs, traces, debug)
                 }
-                e => eyre::bail!("Unrecoverable error: {:?}", e),
+                Err(e) => eyre::bail!("Failed deploying contract: {e:?}"),
```
Comment (Evalir) [inline cli/src/cmd/forge/script/runner.rs:229]: Very nice

File: common/src/traits.rs [hunk 5]
```diff
@@ -67,7 +67,7 @@ impl<'a> TestFunctionExt for &'a str {
     }
 
     fn is_setup(&self) -> bool {
-        self.to_lowercase() == "setup"
+        self.eq_ignore_ascii_case("setup")
```
Comment (Evalir) [inline common/src/traits.rs:70]: very nice catch

File: forge/src/result.rs (hunk 3 above)
Comment (Evalir) [inline forge/src/result.rs:202]: this is really nice!

**Review by Evalir — APPROVED** [review]
this is really nice! love the new `TestSetup` impl

File: forge/src/lib.rs (hunk 1 above)
Comment (mattsse) [reply forge/src/lib.rs:10]: I see, fine with that

**Pull request merged**